src/ui/components/ProtocolViewer.tsx

```diff
@@ -115,49 +135,151 @@ const ProtocolRequestDetail = ({
   );
 };
 
+interface RequestTimings {
+  startTime: number;
+  stopTime: number;
+}
+
 const ProtocolViewer = () => {
-  const requests = useSelector(getProtocolRequests);
-  const chunks = useMemo(() => chunkedRequests(requests), [requests]);
+  const [clearBeforeIndex, setClearBeforeIndex] = useState(0);
+  const [filterText, setFilterText] = useState("");
+  const deferredFilterText = useDeferredValue(filterText);
+
+  const errorMap = useSelector(getProtocolErrorMap);
+  const requestMap = useSelector(getProtocolRequestMap);
+  const responseMap = useSelector(getProtocolResponseMap);
+
   const [selectedChunk, setSelectedChunk] = useState<RequestSummaryChunk | null>(null);
-  const selectedRequestDetails = useSelector(getFullRequestDetails(selectedChunk?.ids ?? []));
+
+  const chunks = useMemo(() => flattenRequests(requestMap, responseMap), [requestMap, responseMap]);
+  const filteredChunks = useMemo(
+    () =>
+      chunks.slice(clearBeforeIndex).filter(chunk => {
+        const fullString = `${chunk.class}.${chunk.method}`.toLowerCase();
+        return fullString.includes(deferredFilterText.toLowerCase());
+      }),
+    [chunks, clearBeforeIndex, deferredFilterText]
+  );
+  const doesFilteredChunksContainSelectedChunk = useMemo(
+    () => selectedChunk !== null && filteredChunks.includes(selectedChunk),
+    [filteredChunks, selectedChunk]
+  );
+
+  const onFilterTextInputChange = (event: React.ChangeEvent) => {
+    setFilterText((event.currentTarget as HTMLInputElement).value);
+  };
+
+  const onClearButtonClick = () => {
```

Comment:
This pattern is striking to me. It seems clearer to have the logic (especially when it is 1 line) embedded on the element itself, rather than having to read the variable assignment, then remember that that variable exists and should be used later, and then try and remember what it does when it gets used eventually, many lines later. I totally get it if the action is complicated, but in that case, it seems like it should not be defined in the component at all. What are your considerations here? When does something go inline vs. defined outside?

Reply:
I generally don't use inline functions because I think they're often less readable. Pulling them out makes it easier to add inline comments (something that's kind of awkward in JSX). It's pretty subjective though.

> and then try and remember what it does when it gets used eventually, many lines later.

IMO clear naming overcomes this, but I guess it's pretty subjective and YMMV.

> What are your considerations here? When does something go inline vs. defined outside?

It's a subjective preference 😄

Reply:
The inline comment thing makes sense. I think the naming would feel more natural if it was like `onClick={clearBefore(index)}`. The pattern of `<button onClick={onButtonClick}>` (or, even worse, `onClick={handleOnClick}` which I find border line maddening), feels like it's not adding much to me.

Reply:
Yeah that's fair. In many cases, `onClick={() => clearBefore(index)}` is fine. When a function does more than 1-2 things, I prefer to pull it out, but it's subjective.